Client/Frontend/Shields/ShieldsViewController.swift

```diff
@@ -0,0 +1,164 @@
+            shieldsView.shieldsContainerStackView.scriptsBlockedStatView.valueLabel.text = "0"
+            shieldsView.shieldsContainerStackView.fingerprintingStatView.valueLabel.text = "0"
+            return
+        }
+        shieldsView.shieldsContainerStackView.adsTrackersStatView.valueLabel.text = String(stats.adsAndTrackers)
+        shieldsView.shieldsContainerStackView.httpsUpgradesStatView.valueLabel.text = String(stats.httpsUpgrades)
+        shieldsView.shieldsContainerStackView.scriptsBlockedStatView.valueLabel.text = String(stats.blockedScripts)
+        shieldsView.shieldsContainerStackView.fingerprintingStatView.valueLabel.text = String(stats.fingerprintingProtection)
+    }
+    
+    private func updateBraveShieldState(shield: BraveShieldState.Shield, on: Bool) {
+        guard let domain = url.normalizedHost else { return }
+        if shield == .AllOff {
+            // Technically we set "all off" when the switch is OFF, unlike all the others
+            BraveShieldState.set(forDomain: domain, state: (.AllOff, !on))
+        } else {
+            BraveShieldState.set(forDomain: domain, state: (shield, on))
+        }
+    }
+    
+    private func updateGlobalShieldState(_ on: Bool, animated: Bool = false) {
+        let updateBlock = {
+            self.shieldsView.shieldsContainerStackView.isHidden = !on
+            self.shieldsView.shieldsContainerStackView.alpha = on ? 1.0 : 0.0
+            self.shieldsView.overviewStackView.isHidden = on
+            self.shieldsView.overviewStackView.alpha = on ? 0.0 : 1.0
+        }
+        if animated {
+            UIView.animate(withDuration: 0.25) {
+                updateBlock()
+            }
+        } else {
+            updateBlock()
+        }
+        shieldsView.stackView.setNeedsLayout()
+        shieldsView.stackView.layoutIfNeeded()
+        preferredContentSize = CGSize(width: 320.0, height: shieldsView.stackView.bounds.height)
+    }
+    
+    // MARK: -
+    
+    /// Groups the shield types with their control and global preference
+    private lazy var shieldControlMapping: [(BraveShieldState.Shield, ToggleView, Preferences.Option<Bool>?)] = [
+        (.AllOff, shieldsView.shieldOverrideControl, nil),
+        (.AdblockAndTp, shieldsView.shieldsContainerStackView.adsTrackersControl, Preferences.Shields.blockAdsAndTracking),
+        (.SafeBrowsing, shieldsView.shieldsContainerStackView.blockMalwareControl, Preferences.Shields.blockPhishingAndMalware),
+        (.NoScript, shieldsView.shieldsContainerStackView.blockScriptsControl, Preferences.Shields.blockScripts),
+        (.HTTPSE, shieldsView.shieldsContainerStackView.httpsUpgradesControl, Preferences.Shields.httpsEverywhere),
+        (.FpProtection, shieldsView.shieldsContainerStackView.fingerprintingControl, Preferences.Shields.fingerprintingProtection),
+    ]
+    
+    var shieldsView: View {
+        return view as! View
+    }
+    
+    override func loadView() {
+        view = View()
```

Comment:
Why do we need to do this manually?

Reply:
Because I am using a custom view (`ShieldsViewController.View`) rather than let `UIViewController` create the view.

Reply:
Maybe I should rename it heh… Using nested classes lets me name it `View` and not worry about collisions but if its too confusing I will just rename it to `ShieldsView`

Reply:
It's harder to see on github, in XCode custom classes are highlighted in different color so it's fine for me.